lib/teleterm/gateway/gateway_test.go

```diff
@@ -92,6 +89,164 @@ func TestGatewayStart(t *testing.T) {
 	require.NoError(t, <-serveErr)
 }
 
+func TestSetLocalPortStartsListenerOnNewPortIfPortIsFree(t *testing.T) {
+	tcpPortAllocator := mockTCPPortAllocator{}
+	gateway := serveGateway(t, &tcpPortAllocator)
+	originalCloseContext := gateway.closeContext
+
+	err := gateway.SetLocalPort("12345")
+	require.NoError(t, err)
+
+	require.Equal(t, "12345", gateway.LocalPort())
+
+	// Verify that the gateway is accepting connections on the new listener.
+	//
+	// mockTCPPortAllocator.Listen returns a listener which occupies a random port but its Addr method
+	// reports the port that was passed to Listen. In order to actually dial the gateway we need to
+	// obtain the real address from the listener.
+	newGatewayAddress := tcpPortAllocator.RecentListener().RealAddr().String()
+	blockUntilGatewayAcceptsConnections(t, newGatewayAddress)
+
+	// Verify that the old context was canceled.
+	//
+	// What we really want to test is if the old listener was closed. Unfortunately, we don't seem to
+	// have a straightforward way to test this as at this point another process might have started
+	// listening on that port.
+	require.ErrorIs(t, originalCloseContext.Err(), context.Canceled,
+		"The listener on the old port wasn't closed after starting a listener on the new port.")
+}
+
+func TestSetLocalPortDoesntStopGatewayIfNewPortIsOccupied(t *testing.T) {
```

Comment:
I could make a table driven test out of those two but is it worthwhile doing this for just two tests?

As someone who started writing Go only after joining Teleport I'm super biased but most of the time the table driven tests in our codebase are far less readable to me than if they were just regular tests with logic shared through regular functions.